### PR TITLE
Add new training pipeline source code with test

### PR DESCRIPTION
**Pull request by sunya-ch**
This PR initially completes the implementation of new pipeline stated in https://github.com/sustainable-computing-io/kepler-model-server/issues/75.

I put new pipeline implementation separately in the new folder to avoid conflict with the current implementation. 
However, when we fully migrates the previous trainer to new pipeline, the old server folder must be replaced.

We can try testing each component of the pipeline as written in workflow action.

```yaml
    - name: Test extractor
      run: |
         python3 tests/extractor_test.py

    - name: Test isolator
      run: |
         python3 tests/isolator_test.py

    - name: Test trainer
      run: |
         python3 tests/trainer_test.py

    - name: Test pipeline
      run: |
         python3 tests/pipeline_test.py
```

The output of `python3 tests/pipeline_test.py` will be in `src/models` folder.
```bash
src/models
├── acpi
│   ├── AbsPower
│   │   ├── BPFOnly
│   │   │   ├── GradientBoostingRegressorTrainer_1
│   │   │   │   ├── GradientBoostingRegressorTrainer_1.json
│   │   │   │   ├── metadata.json
│   │   │   │   ├── platform.pkl
│   │   │   │   └── scaler.pkl
│   │   │   ├── GradientBoostingRegressorTrainer_1.zip
│   │   │   ├── SGDRegressorTrainer_1
│   │   │   │   ├── SGDRegressorTrainer_1.json
│   │   │   │   ├── metadata.json
│   │   │   │   ├── platform.pkl
│   │   │   │   ├── scaler.pkl
│   │   │   │   └── weight.json
│   │   │   ├── SGDRegressorTrainer_1.zip
│   │   │   └── checkpoint
│   │   │       ├── GradientBoostingRegressorTrainer_platform_1.pkl
│   │   │       └── SGDRegressorTrainer_platform_1.pkl
...
│   └── DynPower
│       ├── BPFOnly
│       │   ├── GradientBoostingRegressorTrainer_1
│       │   │   ├── GradientBoostingRegressorTrainer_1.json
│       │   │   ├── metadata.json
│       │   │   ├── platform.pkl
│       │   │   └── scaler.pkl
│       │   ├── GradientBoostingRegressorTrainer_1.zip
│       │   ├── SGDRegressorTrainer_1
│       │   │   ├── SGDRegressorTrainer_1.json
│       │   │   ├── metadata.json
│       │   │   ├── platform.pkl
│       │   │   ├── scaler.pkl
│       │   │   └── weight.json
│       │   ├── SGDRegressorTrainer_1.zip
│       │   └── checkpoint
│       │       ├── GradientBoostingRegressorTrainer_platform_1.pkl
│       │       └── SGDRegressorTrainer_platform_1.pkl
...
```

Note that, the updated server api to use this new pipeline output and document will be pushed by later PRs.

Signed-off-by: Sunyanan Choochotkaew <sunyanan.choochotkaew1@ibm.com>

### CODE REVIEW

File: src/util/train_types.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+###########################################################
+
+import enum
+import random
+
+SYSTEM_FEATURES = ["nodeInfo", "cpu_scaling_frequency_hertz"]
+
+COUNTER_FEAUTRES = ["cache_miss", "cpu_cycles", "cpu_instructions"]
+CGROUP_FEATURES = ["cgroupfs_cpu_usage_us", "cgroupfs_memory_usage_bytes", "cgroupfs_system_cpu_usage_us", "cgroupfs_user_cpu_usage_us"]
+BPF_FEATURES = ["bpf_cpu_time_us", "bpf_block_irq", "bpf_net_rx_irq", "bpf_net_tx_irq"]
```
Comment (KaiyiLiu1234) [inline src/util/train_types.py:18]: note that this file differs a bit from the recently pushed train_types.py (since this file is a bit outdated). If possible include the changes from the more recent train_types in a followup PR (or I can include whichever is easier). Some differences include the inclusion of Sole_IRQ_Features since that is going to be included in dyncomp model. Also the inclusion of a LabelGroup (which I think will help greatly in managing our models.)

Comment (sunya-ch) [reply src/util/train_types.py:18]: Sorry. I will push the update.

File: src/profile/tool/profile_background.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    if not os.path.exists(profile_filename):
+        profile = dict()
+        for component in PowerSourceMap[source]:
+            profile[component] = dict()
+    else:
+        with open(profile_filename) as f:
+            profile = json.load(f)
+    return profile
+
+def save_profile(profile, source):
```
Comment (KaiyiLiu1234) [inline src/profile/tool/profile_background.py:63]: I think we should look at better ways of saving a profile. Maybe we can treat profiles like users (set up a potential db)? This all depends once again on how we plan to manage contributors.

File: src/train/isolator/isolation_trainers/mlp.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    layers.Dense(150, activation='relu', input_shape=[input_dim], kernel_constraint=nonneg()),
+    layers.Dense(1)
+    ])
+    model.compile(loss='mse',
+                optimizer='adam',
+                metrics=['mae', 'mse'])
+    return model
+
+def mlp_2_do(input_dim):
+    model = keras.Sequential([
```
Comment (KaiyiLiu1234) [inline src/train/isolator/isolation_trainers/mlp.py:43]: Looks like a bit of repetetive code? I am pretty sure tensorflow keras with functional api might be better.

File: src/train/trainer/GradientBoostingRegressorTrainer/main.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+from sklearn.ensemble import GradientBoostingRegressor
```
Comment (KaiyiLiu1234) [inline src/train/trainer/GradientBoostingRegressorTrainer/main.py:1]: This should be updated with the XGBoost model in future

Comment (sunya-ch) [reply src/train/trainer/GradientBoostingRegressorTrainer/main.py:1]: Could you update or create new trainer with that?
related issue: https://github.com/sustainable-computing-io/kepler-model-server/issues/89

**Review by KaiyiLiu1234 — APPROVED** [review]
LGTM. For now you can just directly push to kepler-model-db. Final questions I have are what are the scalers and what are their purpose? Why must the isolator trainer be separated from trainer? Is it necessary to have separate models when isolator is implemented? I thought the isolator just cleaned the incoming data (remove idle power I believe?). Please correct me if I am wrong. I can approve the code and any recommendations I made can be addressed later.

File: src/train/isolator/isolation_trainers/mlp.py (hunk 3 above)
Comment (sunya-ch) [reply src/train/isolator/isolation_trainers/mlp.py:43]: Oh... I need to remove it. it is draft code. Thanks.

File: src/profile/tool/profile_background.py (hunk 2 above)
Comment (sunya-ch) [reply src/profile/tool/profile_background.py:63]: That's true. We need the discussion about that to change this design.

**Comment by sunya-ch** [comment]
> LGTM. For now you can just directly push to kepler-model-db. Final questions I have are what are the scalers and what are their purpose? Why must the isolator trainer be separated from trainer? Is it necessary to have separate models when isolator is implemented? I thought the isolator just cleaned the incoming data (remove idle power I believe?). Please correct me if I am wrong. I can approve the code and any recommendations I made can be addressed later.

Because we have to normalize the data before training and we want a scaler that good enough to cover data range (hope to get it from profile). Otherwise, it will just fit to only a batch of training data of the online training. 

Yes, we need to separate when we have to isolator because isolator finds the delta power. The label power is different from the other one.

File: src/util/train_types.py (hunk 1 above)
Comment (sunya-ch) [reply src/util/train_types.py]: @KaiyiLiu1234 I updated only the feature part. Could you please check the rest change related to XGBoost? btw, I think it should not be defined here. 
May related to issue https://github.com/sustainable-computing-io/kepler-model-server/issues/89
